In [ ]:
import pandas as pd
import numpy as np

%cd '/content/drive/MyDrive/Colab Notebooks/NLP HW/HW3/data'

/content/drive/MyDrive/Colab Notebooks/NLP HW/HW3/data


In [ ]:
df = pd.read_table('train', names=['index', 'word_type', 'pos'])
df

index  word_type  pos
0           1     Pierre  NNP
1           2     Vinken  NNP
2           3          ,    ,
3           4         61   CD
4           5      years  NNS
...       ...        ...  ...
912090     22         to   TO
912091     23        San  NNP
912092     24  Francisco  NNP
912093     25    instead   RB
912094     26          .    .

[912095 rows x 3 columns]

In [ ]:
vocabulary = dict()
idx = 0

for row in df.values:
  if row[1] in vocabulary:
    vocabulary[row[1]]['occurrences'] += 1
  else:
    vocabulary[row[1]] = dict()
    vocabulary[row[1]]['index'] = idx
    vocabulary[row[1]]['occurrences'] = 1
    idx += 1

print(vocabulary)
print(len(vocabulary))

{'Pierre': {'index': 0, 'occurrences': 6}, 'Vinken': {'index': 1, 'occurrences': 2}, ',': {'index': 2, 'occurrences': 46476}, '61': {'index': 3, 'occurrences': 25}, 'years': {'index': 4, 'occurrences': 1130}, 'old': {'index': 5, 'occurrences': 213}, 'will': {'index': 6, 'occurrences': 2982}, 'join': {'index': 7, 'occurrences': 41}, 'the': {'index': 8, 'occurrences': 39533}, 'board': {'index': 9, 'occurrences': 297}, 'as': {'index': 10, 'occurrences': 4054}, 'a': {'index': 11, 'occurrences': 18469}, 'nonexecutive': {'index': 12, 'occurrences': 6}, 'director': {'index': 13, 'occurrences': 309}, 'Nov.': {'index': 14, 'occurrences': 238}, '29': {'index': 15, 'occurrences': 74}, '.': {'index': 16, 'occurrences': 37452}, 'Mr.': {'index': 17, 'occurrences': 3856}, 'is': {'index': 18, 'occurrences': 6735}, 'chairman': {'index': 19, 'occurrences': 429}, 'of': {'index': 20, 'occurrences': 22104}, 'Elsevier': {'index': 21, 'occurrences': 1}, 'N.V.': {'index': 22, 'occurrences': 13}, 'Dutch': {'in

In [ ]:
threshold = 3

vocabulary['<unk>'] = dict()
vocabulary['<unk>']['occurrences'] = 0
vocabulary['<unk>']['index'] = -1
rare_words_to_remove = list()

for word in vocabulary:
  if vocabulary[word]['occurrences'] < 3:
    vocabulary['<unk>']['occurrences'] += vocabulary[word]['occurrences']
    rare_words_to_remove.append(word)

for word in rare_words_to_remove:
  vocabulary.pop(word)

print(vocabulary)
print(len(vocabulary))

{'Pierre': {'index': 0, 'occurrences': 6}, ',': {'index': 2, 'occurrences': 46476}, '61': {'index': 3, 'occurrences': 25}, 'years': {'index': 4, 'occurrences': 1130}, 'old': {'index': 5, 'occurrences': 213}, 'will': {'index': 6, 'occurrences': 2982}, 'join': {'index': 7, 'occurrences': 41}, 'the': {'index': 8, 'occurrences': 39533}, 'board': {'index': 9, 'occurrences': 297}, 'as': {'index': 10, 'occurrences': 4054}, 'a': {'index': 11, 'occurrences': 18469}, 'nonexecutive': {'index': 12, 'occurrences': 6}, 'director': {'index': 13, 'occurrences': 309}, 'Nov.': {'index': 14, 'occurrences': 238}, '29': {'index': 15, 'occurrences': 74}, '.': {'index': 16, 'occurrences': 37452}, 'Mr.': {'index': 17, 'occurrences': 3856}, 'is': {'index': 18, 'occurrences': 6735}, 'chairman': {'index': 19, 'occurrences': 429}, 'of': {'index': 20, 'occurrences': 22104}, 'N.V.': {'index': 22, 'occurrences': 13}, 'Dutch': {'index': 23, 'occurrences': 25}, 'publishing': {'index': 24, 'occurrences': 48}, 'group': 

In [ ]:
  #Remove rare words from dataframe
mask = df['word_type'].isin(rare_words_to_remove)
df = df[~mask]
len(df['word_type'].value_counts())

16919

In [ ]:
del idx
del row
del word
del rare_words_to_remove
del threshold
del mask

In [ ]:
print(vocabulary['<unk>']['occurrences'])

32537


In [ ]:
from collections import OrderedDict

sorted_vocabulary = OrderedDict(sorted(vocabulary.items(), key=lambda item: item[1]['occurrences'], reverse=True))

print(sorted_vocabulary)
print(len(sorted_vocabulary))

OrderedDict([(',', {'index': 2, 'occurrences': 46476}), ('the', {'index': 8, 'occurrences': 39533}), ('.', {'index': 16, 'occurrences': 37452}), ('<unk>', {'occurrences': 32537, 'index': -1}), ('of', {'index': 20, 'occurrences': 22104}), ('to', {'index': 46, 'occurrences': 21305}), ('a', {'index': 11, 'occurrences': 18469}), ('and', {'index': 29, 'occurrences': 15346}), ('in', {'index': 97, 'occurrences': 14609}), ("'s", {'index': 110, 'occurrences': 8872}), ('for', {'index': 180, 'occurrences': 7743}), ('that', {'index': 80, 'occurrences': 7723}), ('$', {'index': 477, 'occurrences': 6762}), ('is', {'index': 18, 'occurrences': 6735}), ('``', {'index': 121, 'occurrences': 6673}), ('The', {'index': 67, 'occurrences': 6578}), ("''", {'index': 141, 'occurrences': 6500}), ('said', {'index': 85, 'occurrences': 5418}), ('on', {'index': 148, 'occurrences': 4905}), ('%', {'index': 397, 'occurrences': 4718}), ('it', {'index': 60, 'occurrences': 4509}), ('by', {'index': 244, 'occurrences': 4274})

In [ ]:
with open("vocab.txt", "w") as file:

    file.write(f"<unk>\t{sorted_vocabulary['<unk>']['occurrences']}\n")
    for item, info in sorted_vocabulary.items():
      if item == '<unk>': continue
      file.write(f"{item}\t{info['index']}\t{info['occurrences']}\n")

Q. What is the selected threshold for unknown words replacement?
> Ans - I have chosen value 3 as the threshold for unknown words replacement in the threshold variable.

Q. What is the total size of your vocabulary and what is the total occurrences of the special token '< unk >' after replacement?
> Ans - The total size of the vocabulary after unknown words replacement is 16920
and the occurreneces of special token '< unk >' is 32537

In [ ]:
del file
del info
del item
del vocabulary

In [ ]:
unique_POS_Tags = df.pos.unique().tolist()
unique_POS_Tags = ['Words'] + unique_POS_Tags
print(unique_POS_Tags)

['Words', 'NNP', ',', 'CD', 'NNS', 'JJ', 'MD', 'VB', 'DT', 'NN', 'IN', '.', 'VBZ', 'VBG', 'CC', 'VBD', 'VBN', 'RB', 'TO', 'PRP', 'RBR', 'WDT', 'VBP', 'RP', 'PRP$', 'JJS', 'POS', '``', 'EX', "''", 'WP', ':', 'JJR', 'WRB', '$', 'NNPS', 'WP$', '-LRB-', '-RRB-', 'PDT', 'RBS', 'FW', 'UH', 'SYM', 'LS', '#']


In [ ]:
emission = pd.DataFrame(columns=unique_POS_Tags)
emission

Empty DataFrame
Columns: [Words, NNP, ,, CD, NNS, JJ, MD, VB, DT, NN, IN, ., VBZ, VBG, CC, VBD, VBN, RB, TO, PRP, RBR, WDT, VBP, RP, PRP$, JJS, POS, ``, EX, '', WP, :, JJR, WRB, $, NNPS, WP$, -LRB-, -RRB-, PDT, RBS, FW, UH, SYM, LS, #]
Index: []

[0 rows x 46 columns]

In [ ]:
em_column_names = dict()
i = 0
for col in unique_POS_Tags:
  em_column_names[col] = i
  i += 1

em_row_names = dict()

In [ ]:
# #For testing
i = 0
for row in df.values:
  if row[1] not in em_row_names:
    em_row_names[row[1]] = i
    i += 1

In [ ]:
i = 0
for row in df.values:
  if row[1] in em_row_names:
    emission.iloc[em_row_names[row[1]], em_column_names[row[2]]] += 1
  else:
    em_row_names[row[1]] = i
    i += 1
    temp = pd.Series({'Words': row[1], row[2]: 1})
    emission = pd.concat([emission, temp.to_frame().T], ignore_index=True)

emission.fillna(0, inplace=True)
emission

Words  NNP      ,  CD   NNS   JJ  MD  VB  DT  NN  ...  WP$  -LRB-  \
0          Pierre    6      0   0     0    0   0   0   0   0  ...    0      0   
1               ,    0  46476   0     0    0   0   0   0   0  ...    0      0   
2              61    0      0  25     0    0   0   0   0   0  ...    0      0   
3           years    0      0   0  1130    0   0   0   0   0  ...    0      0   
4             old    0      0   0     0  213   0   0   0   0  ...    0      0   
...           ...  ...    ...  ..   ...  ...  ..  ..  ..  ..  ...  ...    ...   
16914  Steinhardt   34      0   0     0    0   0   0   0   0  ...    0      0   
16915     Granges    2      0   0     0    0   0   0   0   0  ...    0      0   
16916      Tartan    3      0   0     0    0   0   0   0   0  ...    0      0   
16917       Sasea    3      0   0     0    0   0   0   0   0  ...    0      0   
16918     Fiorini    3      0   0     0    0   0   0   0   0  ...    0      0   

       -RRB-  PDT  RBS  FW  UH  SYM  LS  #  
0          0    0    0   0   0    0   0  0  
1          0    0    0   0   0    0   0  0  
2          0    0    0   0   0    0   0  0  
3          0    0    0   0   0    0   0  0  
4          0    0    0   0   0    0   0  0  
...      ...  ...  ...  ..  ..  ...  .. ..  
16914      0    0    0   0   0    0   0  0  
16915      0    0    0   0   0    0   0  0  
16916      0    0    0   0   0    0   0  0  
16917      0    0    0   0   0    0   0  0  
16918      0    0    0   0   0    0   0  0  

[16919 rows x 46 columns]

In [ ]:
del temp
del row
del i
del col

In [ ]:
#Calculating probabilities

for column in emission:
  if column == 'Words': continue

  column_sum = emission[column].sum()
  emission[column] = emission[column].values / column_sum

emission

<ipython-input-24-eed7a2d966d2>:7: RuntimeWarning: divide by zero encountered in divide
  emission[column] = emission[column].values / column_sum
<ipython-input-24-eed7a2d966d2>:7: RuntimeWarning: invalid value encountered in divide
  emission[column] = emission[column].values / column_sum


Words       NNP         ,        CD       NNS        JJ   MD   VB  \
0          Pierre  0.000080  0.000000  0.000000  0.000000  0.000000  0.0  0.0   
1               ,  0.000000  0.999978  0.000000  0.000000  0.000000  0.0  0.0   
2              61  0.000000  0.000000  0.000824  0.000000  0.000000  0.0  0.0   
3           years  0.000000  0.000000  0.000000  0.021572  0.000000  0.0  0.0   
4             old  0.000000  0.000000  0.000000  0.000000  0.004602  0.0  0.0   
...           ...       ...       ...       ...       ...       ...  ...  ...   
16914  Steinhardt  0.000456  0.000000  0.000000  0.000000  0.000000  0.0  0.0   
16915     Granges  0.000027  0.000000  0.000000  0.000000  0.000000  0.0  0.0   
16916      Tartan  0.000040  0.000000  0.000000  0.000000  0.000000  0.0  0.0   
16917       Sasea  0.000040  0.000000  0.000000  0.000000  0.000000  0.0  0.0   
16918     Fiorini  0.000040  0.000000  0.000000  0.000000  0.000000  0.0  0.0   

        DT   NN  ...  WP$  -LRB-  -RRB-  PDT  RBS   FW   UH  SYM   LS    #  
0      0.0  0.0  ...  0.0    0.0    0.0  0.0  NaN  0.0  0.0  0.0  0.0  0.0  
1      0.0  0.0  ...  0.0    0.0    0.0  0.0  NaN  0.0  0.0  0.0  0.0  0.0  
2      0.0  0.0  ...  0.0    0.0    0.0  0.0  NaN  0.0  0.0  0.0  0.0  0.0  
3      0.0  0.0  ...  0.0    0.0    0.0  0.0  NaN  0.0  0.0  0.0  0.0  0.0  
4      0.0  0.0  ...  0.0    0.0    0.0  0.0  NaN  0.0  0.0  0.0  0.0  0.0  
...    ...  ...  ...  ...    ...    ...  ...  ...  ...  ...  ...  ...  ...  
16914  0.0  0.0  ...  0.0    0.0    0.0  0.0  NaN  0.0  0.0  0.0  0.0  0.0  
16915  0.0  0.0  ...  0.0    0.0    0.0  0.0  NaN  0.0  0.0  0.0  0.0  0.0  
16916  0.0  0.0  ...  0.0    0.0    0.0  0.0  NaN  0.0  0.0  0.0  0.0  0.0  
16917  0.0  0.0  ...  0.0    0.0    0.0  0.0  NaN  0.0  0.0  0.0  0.0  0.0  
16918  0.0  0.0  ...  0.0    0.0    0.0  0.0  NaN  0.0  0.0  0.0  0.0  0.0  

[16919 rows x 46 columns]

In [ ]:
emission.fillna(0, inplace=True)

In [ ]:
emission.to_csv('emission.csv', index=False)

In [ ]:
emission = pd.read_csv('emission.csv')
emission

Words       NNP         ,        CD       NNS        JJ   MD   VB  \
0          Pierre  0.000080  0.000000  0.000000  0.000000  0.000000  0.0  0.0   
1               ,  0.000000  0.999978  0.000000  0.000000  0.000000  0.0  0.0   
2              61  0.000000  0.000000  0.000824  0.000000  0.000000  0.0  0.0   
3           years  0.000000  0.000000  0.000000  0.021572  0.000000  0.0  0.0   
4             old  0.000000  0.000000  0.000000  0.000000  0.004602  0.0  0.0   
...           ...       ...       ...       ...       ...       ...  ...  ...   
16914  Steinhardt  0.000456  0.000000  0.000000  0.000000  0.000000  0.0  0.0   
16915     Granges  0.000027  0.000000  0.000000  0.000000  0.000000  0.0  0.0   
16916      Tartan  0.000040  0.000000  0.000000  0.000000  0.000000  0.0  0.0   
16917       Sasea  0.000040  0.000000  0.000000  0.000000  0.000000  0.0  0.0   
16918     Fiorini  0.000040  0.000000  0.000000  0.000000  0.000000  0.0  0.0   

        DT   NN  ...  WP$  -LRB-  -RRB-  PDT  RBS   FW   UH  SYM   LS    #  
0      0.0  0.0  ...  0.0    0.0    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1      0.0  0.0  ...  0.0    0.0    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
2      0.0  0.0  ...  0.0    0.0    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
3      0.0  0.0  ...  0.0    0.0    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
4      0.0  0.0  ...  0.0    0.0    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
...    ...  ...  ...  ...    ...    ...  ...  ...  ...  ...  ...  ...  ...  
16914  0.0  0.0  ...  0.0    0.0    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
16915  0.0  0.0  ...  0.0    0.0    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
16916  0.0  0.0  ...  0.0    0.0    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
16917  0.0  0.0  ...  0.0    0.0    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
16918  0.0  0.0  ...  0.0    0.0    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[16919 rows x 46 columns]

In [ ]:
del file
del column_sum
del idx
del row
del i
del word

Transmission

In [ ]:
trans_column_names = unique_POS_Tags[1:] + ['<EOS>']
trans_row_names = ['<BOS>'] + unique_POS_Tags[1:]

In [ ]:
transmission = pd.DataFrame(columns=trans_column_names)
transmission.insert(0, 'index_col', trans_row_names)
transmission.head()

index_col  NNP    ,   CD  NNS   JJ   MD   VB   DT   NN  ... -LRB- -RRB-  \
0     <BOS>  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...   NaN   NaN   
1       NNP  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...   NaN   NaN   
2         ,  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...   NaN   NaN   
3        CD  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...   NaN   NaN   
4       NNS  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...   NaN   NaN   

   PDT  RBS   FW   UH  SYM   LS    # <EOS>  
0  NaN  NaN  NaN  NaN  NaN  NaN  NaN   NaN  
1  NaN  NaN  NaN  NaN  NaN  NaN  NaN   NaN  
2  NaN  NaN  NaN  NaN  NaN  NaN  NaN   NaN  
3  NaN  NaN  NaN  NaN  NaN  NaN  NaN   NaN  
4  NaN  NaN  NaN  NaN  NaN  NaN  NaN   NaN  

[5 rows x 47 columns]

In [ ]:
trans_col_names = dict()
for idx, col in enumerate(trans_column_names):
  trans_col_names[col] = idx + 1

trans_r_names = dict()
for idx, row in enumerate(trans_row_names):
  trans_r_names[row] = idx

del idx
del col
del row
del trans_column_names
del trans_row_names

In [ ]:
df.reset_index(inplace = True)

In [ ]:
transmission.fillna(0, inplace=True)

In [ ]:
for i in range(0, len(df['pos'])-1):
  if(df['pos'][i] == '.'):
    transmission.iloc[trans_r_names[df['pos'][i]], trans_col_names['<EOS>']] += 1
    transmission.iloc[trans_r_names['<BOS>'], trans_col_names[df['pos'][i+1]]] += 1
    continue

  transmission.iloc[trans_r_names[df['pos'][i]], trans_r_names[df['pos'][i+1]]] += 1

transmission.iloc[trans_r_names[df['pos'][i]], trans_col_names['<EOS>']] += 1

In [ ]:
transmission.head()

index_col    NNP      ,    CD   NNS    JJ    MD   VB    DT    NN  ...  \
0     <BOS>   6647    190   362  1420  1231    43   84  7926  1334  ...   
1       NNP  27962  11237  1638  1985   680   966   87   262  4697  ...   
2         ,   5620    225   945  1276  1656   494  168  6278  2323  ...   
3        CD    396   3000  5369  5024  1136    70    2   728  6621  ...   
4       NNS    168   6732    81   579   978  1546  206  1015  1101  ...   

   -LRB-  -RRB-  PDT  RBS  FW  UH  SYM  LS  #  <EOS>  
0    180    201   26   18   2  15   27  28  0      0  
1    257    284    1    1  36   0    3   0  0      0  
2     18      3   10   17  12  13    0   1  0      0  
3     55    250    0    7   1   0    5   0  0      0  
4    242     90    0    8   0   0    3   0  0      0  

[5 rows x 47 columns]

In [ ]:
#Calculating probabilities for transmission

for column in transmission:
  if column == 'index_col': continue

  column_sum = transmission[column].sum()
  transmission[column] = transmission[column].values / column_sum

transmission.head()

index_col       NNP         ,        CD       NNS        JJ        MD  \
0     <BOS>  0.083660  0.004088  0.011799  0.025999  0.023154  0.004559   
1       NNP  0.351931  0.241760  0.053390  0.036344  0.012790  0.102417   
2         ,  0.070734  0.004841  0.030802  0.023363  0.031148  0.052375   
3        CD  0.004984  0.064544  0.175000  0.091986  0.021367  0.007422   
4       NNS  0.002114  0.144836  0.002640  0.010601  0.018396  0.163910   

         VB        DT        NN  ...     -LRB-     -RRB-       PDT       RBS  \
0  0.003403  0.100621  0.010916  ...  0.137931  0.152157  0.078313  0.041475   
1  0.003524  0.003326  0.038435  ...  0.196935  0.214989  0.003012  0.002304   
2  0.006805  0.079699  0.019009  ...  0.013793  0.002271  0.030120  0.039171   
3  0.000081  0.009242  0.054179  ...  0.042146  0.189251  0.000000  0.016129   
4  0.008345  0.012885  0.009009  ...  0.185441  0.068130  0.000000  0.018433   

         FW        UH   SYM        LS    #  <EOS>  
0  0.013333  0.245902  0.54  0.608696  0.0    0.0  
1  0.240000  0.000000  0.06  0.000000  0.0    0.0  
2  0.080000  0.213115  0.00  0.021739  0.0    0.0  
3  0.006667  0.000000  0.10  0.000000  0.0    0.0  
4  0.000000  0.000000  0.06  0.000000  0.0    0.0  

[5 rows x 47 columns]

In [ ]:
del column
del column_sum

In [ ]:
transmission.to_csv('transmission.csv', index=False)

In [ ]:
transmission = pd.read_csv('transmission.csv')
transmission

index_col       NNP         ,        CD       NNS        JJ        MD  \
0      <BOS>  0.083660  0.004088  0.011799  0.025999  0.023154  0.004559   
1        NNP  0.351931  0.241760  0.053390  0.036344  0.012790  0.102417   
2          ,  0.070734  0.004841  0.030802  0.023363  0.031148  0.052375   
3         CD  0.004984  0.064544  0.175000  0.091986  0.021367  0.007422   
4        NNS  0.002114  0.144836  0.002640  0.010601  0.018396  0.163910   
5         JJ  0.023121  0.036381  0.028944  0.218998  0.069294  0.003817   
6         MD  0.000151  0.000818  0.000065  0.000275  0.000169  0.000106   
7         VB  0.009377  0.009983  0.013070  0.024150  0.037976  0.001484   
8         DT  0.107258  0.010585  0.056551  0.104143  0.293859  0.022371   
9         NN  0.013467  0.304023  0.023403  0.167896  0.019863  0.235263   
10        IN  0.169723  0.017491  0.165678  0.107824  0.150456  0.002969   
11         .  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
12       VBZ  0.014675  0.012672  0.012679  0.006408  0.026973  0.001908   
13       VBG  0.006331  0.004346  0.007008  0.020781  0.016439  0.000742   
14        CC  0.042566  0.007788  0.029824  0.030723  0.042923  0.027460   
15       VBD  0.018917  0.015318  0.054368  0.013219  0.027499  0.001166   
16       VBN  0.005764  0.013425  0.006714  0.012688  0.014822  0.002439   
17        RB  0.002618  0.059725  0.035658  0.003351  0.050691  0.033715   
18        TO  0.011441  0.001807  0.044915  0.011059  0.011869  0.000000   
19       PRP  0.000239  0.007767  0.000456  0.000384  0.002577  0.218087   
20       RBR  0.000013  0.002302  0.000033  0.000128  0.008784  0.000106   
21       WDT  0.001108  0.000538  0.000391  0.001538  0.000903  0.060751   
22       VBP  0.002794  0.006024  0.003553  0.007470  0.017700  0.001590   
23        RP  0.000743  0.002560  0.001630  0.002160  0.002201  0.000106   
24      PRP$  0.004418  0.001312  0.004596  0.027940  0.032220  0.000212   
25       JJS  0.000289  0.001097  0.004400  0.004925  0.003367  0.000318   
26       POS  0.010270  0.001614  0.006030  0.020177  0.029719  0.000954   
27        ``  0.005248  0.001506  0.001695  0.003918  0.012358  0.009436   
28        EX  0.000000  0.000086  0.000000  0.000000  0.000000  0.007103   
29        ''  0.011101  0.000516  0.001108  0.003204  0.002163  0.003923   
30        WP  0.000554  0.000344  0.000130  0.000384  0.000263  0.015903   
31         :  0.006633  0.000538  0.017568  0.003113  0.003273  0.005195   
32       JJR  0.000604  0.002603  0.000196  0.009374  0.002445  0.000318   
33       WRB  0.002542  0.000538  0.000522  0.002710  0.003724  0.002439   
34         $  0.000038  0.003227  0.199609  0.000055  0.000583  0.000212   
35      NNPS  0.008370  0.006304  0.000033  0.000458  0.000301  0.006149   
36       WP$  0.000113  0.000000  0.000065  0.000879  0.000451  0.000000   
37     -LRB-  0.004909  0.000301  0.001239  0.000476  0.000978  0.000212   
38     -RRB-  0.000793  0.004884  0.000326  0.000623  0.000470  0.002757   
39       PDT  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
40       RBS  0.000000  0.000108  0.000000  0.000165  0.005643  0.000000   
41        FW  0.000365  0.000430  0.000000  0.000110  0.000132  0.000106   
42        UH  0.000000  0.000753  0.000000  0.000000  0.000000  0.000000   
43       SYM  0.000025  0.000000  0.000098  0.000000  0.000056  0.000000   
44        LS  0.000000  0.000172  0.000000  0.000000  0.000000  0.000000   
45         #  0.000000  0.000043  0.003814  0.000000  0.000000  0.000000   

          VB        DT        NN  ...     -LRB-     -RRB-       PDT       RBS  \
0   0.003403  0.100621  0.010916  ...  0.137931  0.152157  0.078313  0.041475   
1   0.003524  0.003326  0.038435  ...  0.196935  0.214989  0.003012  0.002304   
2   0.006805  0.079699  0.019009  ...  0.013793  0.002271  0.030120  0.039171   
3   0.000081  0.009242  0.054179  ...  0.042146  0.189251  0.000000  0.016129   
4   0.008345  0.012885  0

In [ ]:
with open("hmm.json", "w") as file:

  #Transmission parameters
  file.write("{\"transmission\":{")
  temp_col_names = list(trans_col_names.keys())
  for idx, row in enumerate(transmission.values):
    for i in range(1, len(row)):
      s = "\'\'" if row[0] == "''" else row[0]
      s_dash = "\'\'" if temp_col_names[i-1] == "''" else temp_col_names[i-1]
      file.write(f'''\"({s_dash},{s})\":
      {transmission.iloc[trans_r_names[s]
          , trans_col_names[s_dash]]}''')
      if(idx+1 == len(transmission.values) and i+1 == len(row)):
        file.write("},")
      else:
        file.write(",")

  #Emission parameters
  file.write("\"emission\":{")
  column_names = emission.columns
  for row in emission.values:
    for i in range(1, len(row)):
      word = "\\\*" if row[0] == "\*" else "\\\*\\\*" if row[0] == "\*\*" else row[0]
      POS = "\'\'" if column_names[i] == "''" else column_names[i]
      file.write(f'''\"({word},{POS})\":
      {emission.iloc[em_row_names[row[0]],
                      em_column_names[column_names[i]]]},''')
  file.write("}}")

In [ ]:
print("The number of transmission parameters: ", transmission.shape[0]*(transmission.shape[1]-1))
print("The number of emission parameters: ", emission.shape[0]*(emission.shape[1]-1))

The number of transmission parameters:  2116
The number of emission parameters:  761355


Q3 Greeedy Decoding

In [54]:
dev_data = pd.read_table('dev', names=['index', 'word_type', 'pos'])
dev_data.head()

index     word_type  pos
0      1           The   DT
1      2       Arizona  NNP
2      3  Corporations  NNP
3      4    Commission  NNP
4      5    authorized  VBD

In [ ]:
em_col_list = list(em_column_names.keys())

In [ ]:
def predictGreedyPOS(prev_pos, curr):
  argmax, pos = 0, ''

  try:
    key = em_row_names[curr]
  except KeyError:
    key = em_row_names['Pierre']

  for idx, e in enumerate(emission.iloc[key, 1:]):
    if e == 0:
      continue

    try:
      row_val = trans_r_names[prev_pos]
    except KeyError:
      row_val = trans_r_names['NNP']

    temp = e * transmission.iloc[row_val, trans_col_names[em_col_list[idx+1]]]
    if temp > argmax:
      argmax = temp
      pos = em_col_list[idx+1]

  return pos

In [ ]:
prev_pos = '<BOS>'
with open("dev_predict", "w") as file:
  for idx, row in enumerate(dev_data.values):
    prev_pos = predictGreedyPOS(prev_pos, row[1])
    file.write(f'{row[0]}\t{row[1]}\t{prev_pos}\n')

    if(idx+1 != len(dev_data.values) and dev_data.values[idx+1][0] == 1):
      file.write('\n')
      prev_pos = '<BOS>'

In [55]:
def eval(pred, gold):
  total = 0
  corr = 0
  try:
      with open(gold, 'r') as gf, open(pred, 'r') as pf, open('outputs.txt', 'w') as offile:
          pf_count = 0
          pf_lines = pf.readlines()
          for gline in gf.readlines():
              gline = gline.strip()
              if len(gline) == 0:

                      # handles missing blank line issue
                      if len(pf_lines[pf_count].strip()) == 0:
                          pf_count+=1
                      continue
              tokens = gline.split()

              # print tokens
              gidx = tokens[0]
              gword = tokens[1]
              glabel = tokens[2]

              pline = pf_lines[pf_count]
              pline = pline.strip()
              tokens = pline.split()

              pword = tokens[1] if len(tokens) > 1 else ''    # handles blank line issue if wrong dev dataset is used
              plabel = tokens[2] if len(tokens) > 2 else ''   # handles missing label issue

              #if gidx != pidx:
              #    print('warning: index mismatch: {}, {}'.format(gidx, pidx))

              #if gword != pword:
              #    print('warning: word mismatch: {}, {}'.format(gword, pword))

              total += 1
              if glabel == plabel:
                  corr += 1

              pf_count += 1

              offile.write(' '.join([gword, pword, glabel, plabel, '\n']))

  except:
      print(repr(gline), repr(pline), total)

  print("total: {}, correct: {}, accuracy: {:.2f}%".format(total, corr, float(corr) * 100 / total))

In [ ]:
eval('dev_predict', 'dev')

total: 131768, correct: 116302, accuracy: 88.26%


In [ ]:
test_data = pd.read_table('test', names=['index', 'word_type'])
prev_pos = '<BOS>'
with open("greedy.out", "w") as file:
  for idx, row in enumerate(test_data.values):
    prev_pos = predictGreedyPOS(prev_pos, row[1])
    file.write(f'{row[0]}\t{row[1]}\t{prev_pos}\n')

    if(idx+1 != len(test_data.values) and test_data.values[idx+1][0] == 1):
      file.write('\n')
      prev_pos = '<BOS>'

Viterbi Decoding Algorithm

In [ ]:
sentenceIndex = list(dev_data[dev_data['index'] == 1].index)

print(sentenceIndex)

[0, 37, 78, 106, 120, 128, 149, 174, 205, 239, 261, 287, 307, 329, 360, 390, 418, 443, 455, 482, 499, 531, 552, 579, 602, 619, 639, 668, 677, 690, 708, 743, 777, 795, 807, 835, 859, 879, 896, 904, 926, 941, 962, 981, 1001, 1033, 1064, 1087, 1100, 1113, 1135, 1162, 1213, 1241, 1263, 1294, 1307, 1322, 1361, 1370, 1403, 1420, 1434, 1443, 1470, 1488, 1510, 1546, 1561, 1587, 1597, 1610, 1646, 1699, 1723, 1758, 1784, 1810, 1827, 1845, 1865, 1901, 1938, 1951, 1981, 2009, 2051, 2072, 2083, 2100, 2123, 2146, 2167, 2199, 2208, 2227, 2252, 2279, 2286, 2302, 2330, 2356, 2394, 2438, 2462, 2490, 2521, 2555, 2588, 2620, 2661, 2696, 2710, 2741, 2770, 2789, 2817, 2835, 2866, 2885, 2912, 2938, 2974, 2988, 3028, 3049, 3090, 3125, 3146, 3176, 3198, 3208, 3236, 3263, 3291, 3321, 3353, 3411, 3424, 3438, 3451, 3485, 3521, 3564, 3578, 3604, 3620, 3631, 3661, 3682, 3694, 3731, 3741, 3767, 3790, 3817, 3876, 3916, 3934, 3972, 4011, 4034, 4039, 4073, 4085, 4118, 4167, 4178, 4216, 4242, 4258, 4296, 4306, 4340, 434

In [ ]:
em_col_list_updated = em_col_list[1:]

In [52]:
k = len(em_col_list_updated)
def predictViterbiPOS(sentence):
  m = len(sentence)
  pos = [''] * m
  piMatrix = np.zeros((m, k))

  try:
    key = em_row_names[sentence[0]]
  except KeyError:
    key = em_row_names['Pierre']

  for i in range(0, k):
    e = emission.iloc[key, i+1]
    t = transmission.iloc[trans_r_names['<BOS>'], trans_col_names[em_col_list_updated[i]]]
    temp = t * e

    if temp > piMatrix[0][i]:
      piMatrix[0][i] = temp
      pos[0] = em_col_list_updated[i]


  for j in range(1, m):
    try:
      key = em_row_names[sentence[j]]
    except KeyError:
      key = em_row_names['Pierre']

    for i in range(0, k):
      for x in range(0, k):
        row_val = trans_r_names[em_col_list_updated[x]]

        e = emission.iloc[key, i+1]
        t = transmission.iloc[row_val, trans_col_names[em_col_list_updated[i]]]
        temp = piMatrix[j-1][x] * t * e

        if temp > piMatrix[j][i]:
          piMatrix[j][i] = temp
          pos[j] = em_col_list_updated[i]

  return pos

In [60]:
with open("dev_viterbi_predict", "w") as file:
  for idx in range(0, len(sentenceIndex)):
    if(idx+1 == len(sentenceIndex)):
      sentence = dev_data.iloc[sentenceIndex[idx]:, 1]
    else:
      sentence = dev_data.iloc[sentenceIndex[idx]:sentenceIndex[idx+1], 1]

    pos = predictViterbiPOS(sentence.tolist())

    for i, word in enumerate(sentence):
      file.write(f'{i+1}\t{word}\t{pos[i]}\n')
    file.write('\n')

In [64]:
eval('dev_viterbi_predict', 'dev')

total: 131768, correct: 114780, accuracy: 87.11%


In [ ]:
test_data = pd.read_table('test', names=['index', 'word_type'])
sentenceIndex = list(test_data[test_data['index'] == 1].index)

print(len(sentenceIndex))

5462


In [ ]:
with open("viterbi.out", "w") as file:
  for idx in range(0, len(sentenceIndex)):
    if(idx+1 == len(sentenceIndex)):
      sentence = test_data.iloc[sentenceIndex[idx]:, 1]
    else:
      sentence = test_data.iloc[sentenceIndex[idx]:sentenceIndex[idx+1], 1]

    pos = predictViterbiPOS(sentence.tolist())

    for i, word in enumerate(sentence):
      file.write(f'{i+1}\t{word}\t{pos[i]}\n')
    file.write('\n')